File: cookbooks/from-sql-to-arcflow/03-multi-hop-industry.py
"""
Question 3: "Find every person within 2 hops of Alice (via shared employers)
who has worked in biotech."

This is the question SQL handles worst: the path itself is the data, not
just the endpoints. SQL needs to enumerate intermediate nodes via repeated
self-joins; the depth is hardcoded in the SQL. Cypher walks the path.

Real-world version: "show me every analyst within 2 reporting hops of the
fraud incident" / "show me every codebase symbol within 3 import hops of
the changed file" — same shape, different domain.
"""

from _common import make_duckdb, make_arcflow

print("=" * 68)
print("Q3: People within 2 employment-hops of Alice who've worked in biotech")
print("=" * 68)

# ──────────────────────── DuckDB / SQL ────────────────────────
print("\n--- DuckDB / SQL ---")
sql = """
WITH alice AS (
    SELECT id FROM persons WHERE name = 'Alice'
),
alice_orgs AS (                          -- hop 1: Alice's orgs
    SELECT e.org_id FROM employment e JOIN alice ON e.person_id = alice.id
),
hop1_people AS (                         -- people who shared an org with Alice
    SELECT DISTINCT e.person_id
    FROM employment e
    JOIN alice_orgs ao ON e.org_id = ao.org_id
    WHERE NOT e.person_id = (SELECT id FROM alice)
),
hop2_orgs AS (                           -- hop 2: orgs of those people
    SELECT DISTINCT e.org_id
    FROM employment e
    JOIN hop1_people hp ON e.person_id = hp.person_id
)
SELECT DISTINCT p.name, o.name AS via_org
FROM employment e
JOIN persons p ON p.id = e.person_id
JOIN orgs o ON o.id = e.org_id
WHERE e.org_id IN (SELECT org_id FROM hop2_orgs)
  AND o.industry = 'biotech'
  AND p.id <> (SELECT id FROM alice)
ORDER BY p.name
"""
print(sql.strip())
print("\nresult:")
db = make_duckdb()
for r in db.execute(sql).fetchall():
    print(f"  {r[0]!r:>10} via {r[1]!r}")
db.close()

# ──────────────────────── ArcFlow / Cypher ────────────────────────
# arcflow-core#13 (lexer) resolved 2026-05-05 (oz-arcflow >= 1.6.9): `!=`
# is now an alias for `<>`. NOT x = y is kept here only because the
# multi-anchor predicates below trip the still-open arcflow-core#10 +
# arcflow-core#14 planner bugs.
# FIXME(arcflow-core#14): inline {industry: 'biotech'} on terminal anchor
#   is silently ignored on multi-hop patterns. The natural form should be:
#     ...->(p:Person)-[:EMPLOYED_AT]->(o2:Org {industry: 'biotech'})
# FIXME(arcflow-core#10): WHERE clause combining multi-anchor predicates
#   returns 0 rows. So the natural WHERE form also fails:
#     WHERE NOT p.name = 'Alice' AND o2.industry = 'biotech'
# Until both ship, the working pattern is: pull all hops, filter in app.
print("\n--- ArcFlow / Cypher (with Python post-filter) ---")
cypher = """
MATCH (alice:Person {name: 'Alice'})-[:EMPLOYED_AT]->(o1:Org)<-[:EMPLOYED_AT]-(p:Person)-[:EMPLOYED_AT]->(o2:Org)
WHERE NOT p.name = 'Alice'
RETURN DISTINCT p.name AS name, o2.name AS via_org, o2.industry AS industry
ORDER BY name
"""
print(cypher.strip())
print("\nresult (biotech-filtered in Python until #10/#14 ship):")
af = make_arcflow()
results = [r for r in af.execute(cypher) if r["industry"] == "biotech"]
for r in results:
    print(f"  {r['name']!r:>10} via {r['via_org']!r}")
af.close()

print("\n--- Observation ---")
print("SQL: 5 CTEs, 4 joins, hardcoded path depth. To extend to 3 hops, add")
print("hop3_orgs and another join chain. Each hop multiplies the join count.")
print("")
print("Cypher: one MATCH that walks the path. To extend to 3 hops, add")
print("`(p)-[:EMPLOYED_AT]->(o3:Org)` and one more variable. The graph engine")
print("plans the traversal; you don't write the join order.")

File: examples/from-sql-to-arcflow/rust/src/bin/03-multi-hop-industry.rs
//! Step 03 (Rust) — multi-hop industry traversal, both engines.
//!
//! Mirrors `03-multi-hop-industry.py`. The graph form walks two hops
//! in one MATCH; the SQL form needs five CTEs and a hard-coded depth.

use anyhow::Result;
use from_sql_to_arcflow::load_both;

fn main() -> Result<()> {
    let both = load_both()?;

    println!("=== People within 2 hops of Alice who've worked in biotech ===\n");

    println!("--- SQL (DuckDB) — 5 CTEs, hard-coded depth ---");
    let sql_q = "
        WITH alice AS (
            SELECT id FROM persons WHERE name = 'Alice'
        ),
        alice_orgs AS (
            SELECT DISTINCT e.org_id FROM employment e, alice
            WHERE e.person_id = alice.id
        ),
        coworkers AS (
            SELECT DISTINCT e.person_id FROM employment e, alice_orgs
            WHERE e.org_id = alice_orgs.org_id
        ),
        coworker_orgs AS (
            SELECT DISTINCT e.org_id FROM employment e, coworkers
            WHERE e.person_id = coworkers.person_id
        ),
        biotech_orgs AS (
            SELECT id FROM orgs WHERE industry = 'biotech'
        )
        SELECT DISTINCT p.name
        FROM employment e
        JOIN persons p ON p.id = e.person_id
        JOIN coworker_orgs ON coworker_orgs.org_id = e.org_id
        JOIN biotech_orgs  ON biotech_orgs.id = e.org_id
        WHERE p.name != 'Alice'
        ORDER BY p.name
    ";
    println!("{}", sql_q.trim());
    let mut stmt = both.sql.prepare(sql_q)?;
    let rows: Vec<String> = stmt
        .query_map([], |row| row.get::<_, String>(0))?
        .filter_map(|r| r.ok())
        .collect();
    println!("→ {rows:?}\n");

    println!("--- WorldCypher (ArcFlow) — one MATCH chain ---");
    let cy = "MATCH (alice:Person {name: 'Alice'})-[:WORKED_AT]->(:Org)<-[:WORKED_AT]-\
              (coworker:Person)-[:WORKED_AT]->(o:Org {industry: 'biotech'}) \
              WHERE coworker.name <> 'Alice' \
              RETURN DISTINCT coworker.name AS name \
              ORDER BY name";
    println!("{cy}");
    let result = both.graph.execute(cy)?;
    let names: Vec<String> = result.rows.into_iter()
        .filter_map(|r| r.get("name").cloned())
        .collect();
    println!("→ {names:?}\n");

    println!("Verdict: each extra hop adds one anchor in Cypher; one more CTE in SQL.");
    Ok(())
}
